frb/analysis/dmhalos_photoz.py

```diff
@@ -0,0 +1,615 @@
+
+    # A similar column for separation in kpc
+    #trim_tab['sep_kpc'] = p15.angular_diameter_distance(trim_tab['redshift']).to('kpc').value*trim_tab['sep_ang']*u.arcmin.to('rad')
+
+    # Rename the stellar mass columns
+    trim_tab.rename_columns(['bayes.stellar.m_star', 'bayes.stellar.m_star_err'],['log_mstar', 'log_mstar_err'])
+    # Convert to logarithmic values
+    trim_tab['log_mstar_err'] = (np.log10(trim_tab['log_mstar']+trim_tab['log_mstar_err']) - 
+                                np.log10(np.abs(trim_tab['log_mstar']-trim_tab['log_mstar_err'])))/2
+    trim_tab['log_mstar'] = np.log10(trim_tab['log_mstar'])
+
+    return trim_tab
+
+def _sample_eazy_redshifts(gal_ID:int, eazy_outdir:str, ndraws:int = 1000)->np.ndarray:
+    """
+    Returns a sample of redshifts drawn from the
+    EAZY photo-z PDF of galaxy <gal_iD>.
+    Args:
+        gal_ID(int): ID number of the galaxy in the EAZY table.
+        eazy_outdir(str): Path to the EAZY results folder
+        ndraws(int, optional): Number of redshift samples desired.
+    Returns:
+        sample_z (np.ndarray): Redshift sample array of length ndraws.
+    """
+    # Get posterior
+    zgrid, pz = ez.getEazyPz(gal_ID-1,OUTPUT_DIRECTORY=eazy_outdir)
+    # Force a value of 0 at z = 0
+    zgrid = np.hstack([[0],zgrid])
+    pz = np.hstack([[0],pz])
+    if np.all(np.diff(zgrid) == 0):
+        return -99
+
+    # make a CDF
+    cdf_z = np.cumsum(pz)
+    cdf_z /= np.max(cdf_z)
+    cdf_interp = interp1d(cdf_z, zgrid, kind="linear", fill_value=0, bounds_error=False)
+
+    # Use uniform distribution to produce random draws from the CDF
+    sample_u = np.random.rand(ndraws)
+    sample_z = cdf_interp(sample_u)
+    return sample_z
+
+def _mhalo_lookup_table(z:float, fits_out:str = "m_halo_realizations.fits"):
+    """
+    For a given z, produce realizations of m_halo for relevant
+    m_star values using only the uncertainty in the SHMR relation.
+    Internal function. Use directly if you know what you're doing.
+    Args:
+        z (float): redshift
+        fits_out(str, optional): output file path.
+    """
+
+    # Define a range of stellar masses
+    n_star = 1000
+    log_mstar_array = np.linspace(6, 11, n_star)
+
+    # Instantiate a 2D array
+    n_halo = 10000
+    log_mhalo_array = np.zeros((n_star, n_halo))
+
+    def mhalo_factory(log_mstar:float, z:float, n_cores = 8)->np.ndarray:
+        """
+        Parallelize m_halo computations for a given log_mstar array.
+        """    
+        p = Pool(n_cores)
+        func = lambda x: halomass_from_stellarmass(x, z = z, randomize=True)
+        log_mhalo_array = p.map(func, log_mstar)
+        
+        return log_mhalo_array
+
+
+    # Loop over log_mstar:
+    for idx, log_mstar in enumerate(log_mstar_array):
+        temp_log_mstar = np.full(n_halo, log_mstar)
+        log_mhalo_array[idx] = mhalo_factory(temp_log_mstar, z = z, n_cores = 10)
+    
+    # Store this in a fits file
+    prihdu  = fits.PrimaryHDU()
```

Comment:
I tend to prefer `numpy` or `hdf5` files these
days unless the extras (e.g. Header) of FITS
is warranted.

Reply:
Noted and changed `.fits` to `.npz` wherever possible.